NUMBERS-156: replacing SafeNorm with Norms and Summation

darkma773r opened the pull request:
- replacing SafeNorm with Norms
- adding Summation class
- adding norm algorithm evaluation tests and benchmarks

Note that this change requires JDK 9+ for examples-jmh and commons-numbers-arrays unit tests.

darkma773r commented on commons-numbers-arrays/src/main/java/org/apache/commons/numbers/arrays/Norms.java:
```diff
@@ -0,0 +1,427 @@
+        if (max > LARGE_THRESH) {
+            scale = SCALE_DOWN;
+            rescale = SCALE_UP;
+        } else if (max < SAFE_SCALE_UP_THRESH) {
```
@aherbert, this is slightly different than the implementation in the ExtendedPrecisionLinearCombination algorithm in examples-jmh. In that class, we scale up if the max is under a threshold for safely scaling up _and_ the min value is less than the threshold for very small numbers. My thought here is that we only need to check the max. If the max value is small enough that we can scale up without overflow, then we may as well just go ahead and do it and not bother checking (or even computing) the min. Let me know if you agree with this assessment.

aherbert replied:
The logic makes sense since you only have two numbers. Note you can change this to a new value:
```java
SAFE_SCALE_UP_THRESH = 0x1.0p-100;
```
You are scaling by 2^600. So this value scaled would be 2^500 and safe against overflow when squared.

With two numbers you do not need to do the computation if the difference between them is large because max^2 + min^2 is mute:
```java
if (max - min > 0x1.0p54) {
    return max;
}
```
So knowing min is useful for a fast exit.

I'll do a full review of this later. But the thresholds should be changed to 2^496 from 2^500 for the array case.

darkma773r replied:
Why the change for the array case? Also, should/could that apply to the other cases as well?

aherbert replied:
Because if you have a lot of memory you can pass in an array of 2^31-1 doubles. If these all have a value of Double.MAX_VALUE then you will overflow during the sum unless you lower the threshold. When you only have 2 or 3 doubles then you could make the threshold higher.

I think you need 2^496 for the array case and 2^510 for the case of 3 inputs. This would be 2^992 or 2^1020 when squared. You can add respectively 2^31 of the former and 3 of the latter.

aherbert replied:
I should add that due to the extended precision summation you do not want it to sum to infinite. This will lead to errors in the round-off computation that creates NaN. So even if the final result would be infinite you do not want it to return NaN. So the thresholds should allow the summation to complete to a finite value. The final scaling can then create the infinite if the result is not possible within the range of a double.

darkma773r replied:
I've updated the threshold and added the early exit check for Euclidean 2D. I don't see a compelling reason to use two separate thresholds, so all Euclidean methods use the 2^496 threshold. The early exit in the 2D case gave a small improvement in performance in the JMH benchmark, small enough where the outliers of the before and after case overlap.

darkma773r replied:
Let me know if you see any other changes that should be made.

darkma773r replied:
@aherbert, did you still want to review this or do you think it's ready to go?

aherbert replied:
Sorry, I have been busy. I will try and find time this weekend.

darkma773r replied:
No worries! I was just making sure there wasn't something I was missing here.

aherbert commented on commons-numbers-arrays/src/main/java/org/apache/commons/numbers/arrays/Norms.java:
```diff
@@ -0,0 +1,451 @@
+    }
+
+    /** Compute the maximum norm (also known as the infinity norm or L<sub>inf</sub> norm) of the given values.
+     * The result is equal to \(\max{(|v_0|, ... |v_i|)}\), i.e., the maximum of the absolute values of the
```
This should be `\(\max{(|v_0| + ... + |v_i|)}\)` to match the style of the other javadocs.

I rendered the javadocs and prefer the `\ldots`. Also use the length of the input n so the summation is to n-1.

Edit to `\(\max{(|v_0| +  \ldots +  |v_{n-1}|)}\)`.

```
rm -f target/maven-javadoc-plugin-stale-data.txt
mvn javadoc:javadoc
open target/site/apidocs/org/apache/commons/numbers/arrays/Norms.html
```

darkma773r replied:
I can't use '+' since that would produce the manhattan norm. I'll update the subscript and ellipsis, though.

aherbert commented on commons-numbers-arrays/src/test/java/org/apache/commons/numbers/arrays/DoubleTestUtils.java:
```diff
@@ -0,0 +1,70 @@
+        final long mask = ((1L << 52) - 1) | 1L << 63;
+        final long bits = rng.nextLong() & mask;
+        // The exponent must be unsigned so + 1023 to the signed exponent
+        final int expRange = Math.abs(maxExp - minExp);
```
Should this be Math.abs(maxExp - minExp) + 1?

Otherwise the call to rng.nextInt(int) could be called with 0 when max is the same as min.

darkma773r replied:
Yep. Oops.

aherbert commented on commons-numbers-arrays/src/test/java/org/apache/commons/numbers/arrays/ExtendedPrecisionTest.java:
```diff
@@ -102,6 +102,23 @@ void testSubNormalSplit() {
         final double lo2 = a - hi2;
         Assertions.assertEquals(a, hi2);
         Assertions.assertEquals(0, lo2);
+
```
Remove extra line

aherbert commented on commons-numbers-arrays/src/test/java/org/apache/commons/numbers/arrays/ExtendedPrecisionTest.java:
```diff
@@ -102,6 +102,23 @@ void testSubNormalSplit() {
+    void testSquareLowUnscaled() {
+        assertSquareLowUnscaled(0.0, 1.0);
+        assertSquareLowUnscaled(0.0, -1.0);
+        assertSquareLowUnscaled(-1.4293872661474477E-16, Math.PI);
```
Can you compute this result using BigDecimal?
```java
final BigDecimal pi = new BigDecimal(Math.PI);
final double pi2low = pi.pow(2).subtract(new BigDecimal(Math.PI * Math.PI)).doubleValue();
assertSquareLowUnscaled(pi2low, Math.PI);
```

It is a bit cryptic to have a constant here.

darkma773r replied:
That sounds good but the expected value computed with `BigDecimal` is `-8.79470486217151E-16` whereas the `ExtendedPrecision` result is '-1.4293872661474477E-16'. Is this simply a limitation of the algorithm or is this a bug in our implementation?

aherbert replied:
I thought the BigDecimal value was the same. I just tried the code above on a different machine and it still works. Stepping through a debugger the value of `pi2low` is `-1.429...`. Are you using a rounding mode for BigDecimal?

aherbert replied:
Since the code is on Java 9 for testing you can drop BigDecimal and simplify this to:
```java
assertSquareLowUnscaled(Math.fma(Math.PI, Math.PI, -Math.PI * Math.PI), Math.PI);
```

darkma773r replied:
I found why `BigDecimal` wasn't working for me: I was using `BigDecimal.valueOf()` to create instances instead of `new BigDecimal()`. Switching to the constructor to use the exact double representation fixed the issue. I'm switching to the `fma` version, however since that is simpler.

aherbert commented on commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/DoubleUtils.java:
```diff
@@ -0,0 +1,72 @@
+        final long mask = ((1L << 52) - 1) | 1L << 63;
+        final long bits = rng.nextLong() & mask;
+        // The exponent must be unsigned so + 1023 to the signed exponent
+        final long exp = rng.nextInt(Math.abs(maxExp - minExp)) + minExp + 1023;
```
Again the range should have a +1 to avoid calling nextInt with the wrong limit.

aherbert commented on commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/DoubleUtils.java:
```diff
@@ -0,0 +1,72 @@
+
+/** Class containing utility methods for working with doubles.
+ */
+public final class DoubleUtils {
```
Does all this have to be public? I think it can move to the arrays package and be package private.

darkma773r replied:
Yep. Done.

aherbert commented on commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/arrays/EuclideanNormAlgorithmPerformance.java:
```diff
@@ -0,0 +1,235 @@
+
+    /** String indicating double exponents over a very wide range of values. */
+    private static final String FULL = "full";
+
```
Remove the extra line

aherbert commented on commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/arrays/EuclideanNormAlgorithmPerformance.java:
```diff
@@ -0,0 +1,235 @@
+public class EuclideanNormAlgorithmPerformance {
+
+    /** Default number of samples used in each benchmark. */
+    private static final int DEFAULT_SAMPLES = 100_000;
```
This is only used in one place. Just hard code it there and avoid the ugly `int + ""` declaration in the `@Param` tag.

Same for DEFAULT_VECTOR_LEN.

aherbert commented on commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/arrays/EuclideanNormAlgorithmPerformance.java:
```diff
@@ -0,0 +1,235 @@
+            default:
+                minExp = -550;
+                maxExp = +550;
+                break;
```
I prefer here to throw in the default case statement. So if you pass parameters to JMH on the command line it will fail:
```
mvn package -Pexamples-jmh
java -jar target/examples-jmh.jar EuclideanNormAlgorithmPerformance -p vectorLength=10 -p type=foobar
```

This should error but instead runs with `type=FULL`.

aherbert commented on commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/arrays/EuclideanNormEvaluator.java:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    /** Compute the ulp difference between two values. This method assumes that
+     * the arguments have the same exponent.
```
They do not have to have the same exponent. The ulp difference between 1.0 and Math.nextDown(1.0) will be 1. They do have to have the same sign. All your results should be positive so this is OK. 

If the numbers are very different then this will return a truncated difference by long to int primitive conversion. A return type of `long` would fix this. I do not think that you require `int` during testing as the sums use a double.

This comment also applies to `DoubleTestUtils` in the main package.

aherbert commented on commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/arrays/EuclideanNormEvaluator.java:
```diff
@@ -0,0 +1,248 @@
+                }
+            }
+
+            final double stdDev = Math.sqrt(diffSumSq / (successCount - 1));
```
In the ridiculous case where success count is 1 this will be invalid. zero successes will compute a NaN mean which is a useful marker, and stdDev will be zero.

aherbert commented on commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/arrays/NormsPerformance.java:
```diff
@@ -0,0 +1,194 @@
+            for (int i = 0; i < vectors.length; ++i) {
+                // pick a general range for the vector element exponents and then
+                // create values within that range
+                final int vMidExp = rng.nextInt(MAX_EXP - MIN_EXP) + MIN_EXP;
```
Invalid range again. Add a 1 to it.

aherbert commented on commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/arrays/NormsPerformance.java:
```diff
@@ -0,0 +1,194 @@
+public class NormsPerformance {
+
+    /** Number of samples used in each benchmark. */
+    private static final int SAMPLES = 100_000;
```
Can all these be put into `@Param` annotated attributes in VectorArrayInput. This allows them to be adjusted by passing values to JMH on the command line.

aherbert commented on commons-numbers-examples/examples-jmh/src/test/java/org/apache/commons/numbers/examples/jmh/arrays/EuclideanNormAccuracyTest.java:
```diff
@@ -0,0 +1,142 @@
+        }
+    }
+
+    /** Create an array of random input vectors with exponents in the range {@code [minExp, maxExp]}..
```
Two periods at the end of the line.

aherbert commented on commons-numbers-arrays/src/test/java/org/apache/commons/numbers/arrays/NormsTest.java:
```diff
@@ -0,0 +1,564 @@
+        final double normResult = computeUlpErrorStdDev(inputs, exactResults, v -> Norms.euclidean(v[0], v[1]));
+
+        // ensure that we are within the ballpark of Math.hypot
+        Assertions.assertTrue(normResult <= (hypotResult + HYPOT_COMPARE_EPS),
```
You should also check the mean. What if Norms computed the exact result but shifted by 20 ULP. It would have a standard deviation of zero but a mean of 20. It would be very precise in its inaccuracy. Effectively you should test accuracy (mean ulp) and precision (std dev of the ulp).

aherbert reviewed (changes requested):
The code looks good. The `Norms.euclidean(double, double)` should be fixed to correctly detect NaN as intended, not by computing and then returning NaN. I have added a few notes on minor nits which you can fix as you see fit.

aherbert commented:
One further note is that this change seems to break coveralls reporting. The current master is working fine. You can see the coveralls error in the Travis log. I am not sure why but this can be investigated separately. It may be due to the use of JDK 9 for travis. Perhaps bump it to JDK 11.

aherbert commented:
The problems with coveralls has been reported elsewhere for JDK 9, e.g:
[coveralls-maven-plugin fails on Java 9](https://github.com/trautonen/coveralls-maven-plugin/issues/112)

This issue is now closed via a merge so perhaps update the coveralls plugin version in the POM to the latest and see if we get a coverage report.

darkma773r commented on commons-numbers-arrays/src/main/java/org/apache/commons/numbers/arrays/Norms.java:
```diff
@@ -0,0 +1,451 @@
+        final double yabs = Math.abs(y);
+        final double zabs = Math.abs(z);
+
+        final double max = Math.max(Math.max(xabs, yabs), zabs);
```
The absolute values are used if I use the above fiddle of returning `xabs * yabs * zabs` as mentioned above, which I quite like.

darkma773r commented on commons-numbers-arrays/src/main/java/org/apache/commons/numbers/arrays/Norms.java:
```diff
@@ -0,0 +1,451 @@
+
+        // if the max is not finite, then one of the inputs must not have
+        // been finite
+        if (!Double.isFinite(max)) {
```
I went with the simple version.

darkma773r commented on commons-numbers-arrays/src/main/java/org/apache/commons/numbers/arrays/Norms.java:
```diff
@@ -0,0 +1,451 @@
+        // if the max is not finite, then one of the inputs must not have
+        // been finite
+        if (!Double.isFinite(max)) {
+            if (Double.isNaN(x) || Double.isNaN(y) || Double.isNaN(z)) {
```
I like it!

darkma773r commented:
@aherbert, thanks for the thorough review! I've implemented the changes you requested except as noted in the comments. Coveralls is still failing, even on JDK 11. There does not appear to be a later version to upgrade to, unfortunately. Let me know if any other changes need to be made.

Note that I also added an exception to the spotbugs exclude file for the new floating point equality expression in Angle.Normalization. The build was failing otherwise.

aherbert commented on commons-numbers-arrays/src/main/java/org/apache/commons/numbers/arrays/ExtendedPrecision.java:
```diff
@@ -204,6 +204,28 @@ private static double productLowUnscaled(double x, double y, double xy) {
+     * finite numbers can create NaN results.
+     *
+     * @param x Number to square
+     * @param xx Standard precision product {@code x*x)
```
Javadoc issue: This should have a `}` not a `)`.

aherbert commented on commons-numbers-arrays/src/main/java/org/apache/commons/numbers/arrays/ExtendedPrecision.java:
```diff
@@ -204,6 +204,28 @@ private static double productLowUnscaled(double x, double y, double xy) {
+     * @param x Number to square
+     * @param xx Standard precision product {@code x*x)
+     * @return the low part of the square double length number
+     * @see #productLowUnscaled(double, double, double)
```
Javadoc issue: The productLowUnscaled is private. So you cannot refer to it in a package private method.

aherbert reviewed (commented):
Without coveralls I ran jacoco locally and found 100% coverage. This is good. My IDE picked up a javadoc error in a new method in ExtendedPrecision.

Could you try this fix for coveralls. Add this to the coveralls configuration in the travis profile:
```xml
<version>${commons.coveralls.version}</version>
<!-- Workaround for coveralls plugin JDK 9 compatibility issue.
     https://github.com/trautonen/coveralls-maven-plugin/issues/112 -->
<dependencies>
    <dependency>
        <groupId>javax.xml.bind</groupId>
        <artifactId>jaxb-api</artifactId>
        <version>2.2.3</version>
    </dependency>
</dependencies> 
```

darkma773r commented on commons-numbers-examples/examples-jmh/src/main/java/org/apache/commons/numbers/examples/jmh/arrays/DoubleUtils.java:
```diff
@@ -0,0 +1,72 @@
+        final long mask = ((1L << 52) - 1) | 1L << 63;
+        final long bits = rng.nextLong() & mask;
+        // The exponent must be unsigned so + 1023 to the signed exponent
+        final long exp = rng.nextInt(Math.abs(maxExp - minExp) + 1) + minExp + 1023;
```
I just documented the requirement.

darkma773r commented:
The build is passing and it looks like coveralls is working as well, although it's not showing up here in the PR for some reason. Any more changes or do you think we're good to go?

aherbert commented:
I see that the latest travis build did create a [coveralls report](https://coveralls.io/jobs/81536135). So perhaps it just takes a while to filter through and e-mail the results.

I think this is OK to merge. If you do that make sure to update the change.xml.

Thanks for the effort.

darkma773r commented:
Thanks! I'll try to merge this tonight and update the changes.xml.
